pkg/controller/upload-controller.go

```diff
@@ -490,37 +332,320 @@ func (c *UploadController) getOrCreateUploadService(pvc *v1.PersistentVolumeClai
+		}
+		if err := r.Client.Create(context.TODO(), pod); err != nil {
+			return nil, err
+		}
+	}
+
+	r.Log.V(1).Info("upload pod created\n", "Namespace", pod.Namespace, "Name", pod.Name, "Image name", args.Image)
+	return pod, nil
+}
+
+// NewUploadController creates a new instance of the upload controller.
+func NewUploadController(mgr manager.Manager, cdiClient *cdiclientset.Clientset, k8sClient kubernetes.Interface, log logr.Logger, uploadImage, pullPolicy, verbose string, serverCertGenerator generator.CertGenerator, clientCAFetcher fetcher.CertBundleFetcher) (controller.Controller, error) {
+	reconciler := &UploadReconciler{
+		Client:              mgr.GetClient(),
+		Scheme:              mgr.GetScheme(),
+		CdiClient:           cdiClient,
+		K8sClient:           k8sClient,
+		Log:                 log.WithName("upload-controller"),
+		Image:               uploadImage,
+		Verbose:             verbose,
+		PullPolicy:          pullPolicy,
+		recorder:            mgr.GetEventRecorderFor("upload-controller"),
+		serverCertGenerator: serverCertGenerator,
+		clientCAFetcher:     clientCAFetcher,
+	}
+	uploadController, err := controller.New("upload-controller", mgr, controller.Options{
+		Reconciler: reconciler,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if err := addUploadControllerWatches(mgr, uploadController); err != nil {
+		return nil, err
+	}
+
+	return uploadController, nil
+}
+
+func addUploadControllerWatches(mgr manager.Manager, importController controller.Controller) error {
+	// Setup watches
+	if err := importController.Watch(&source.Kind{Type: &corev1.PersistentVolumeClaim{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+	if err := importController.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
+		OwnerType:    &corev1.PersistentVolumeClaim{},
+		IsController: true,
+	}); err != nil {
+		return err
+	}
+	if err := importController.Watch(&source.Kind{Type: &corev1.Service{}}, &handler.EnqueueRequestForOwner{
+		OwnerType:    &corev1.PersistentVolumeClaim{},
+		IsController: true,
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// getUploadResourceName returns the name given to upload resources
+func getUploadResourceName(name string) string {
+	// TODO revisit naming, could overflow
+	return "cdi-upload-" + name
+}
+
+// UploadPossibleForPVC is called by the api server to see whether to return an upload token
+func UploadPossibleForPVC(pvc *v1.PersistentVolumeClaim) error {
+	if _, ok := pvc.Annotations[AnnUploadRequest]; !ok {
+		return errors.Errorf("PVC %s is not an upload target", pvc.Name)
+	}
+	return nil
+}
+
+// GetUploadServerURL returns the url the proxy should post to for a particular pvc
+func GetUploadServerURL(namespace, pvc, uploadPath string) string {
+	return fmt.Sprintf("https://%s.%s.svc%s", getUploadResourceName(pvc), namespace, uploadPath)
+}
+
+func makeUploadPodSpec(image, verbose, pullPolicy, name string,
```

Comment:
why is `MakeUploadServiceSpec` a member and this is not?  Symmetry would be nice.  Can also get rid of some args.

Reply:
Good point, this came out of the util.go file, and I just didn't make it a member.